### ISSUE TITLE
Upgrade pybind11 to 2.2+

### ISSUE THREAD

**Issue by qianyizh** [IntelVCL/Open3D#18]
The current pybind11 is modified from 2.0 dev.
The main modifications are:

1. Direct (reference) access of Eigen matrices, using this patch https://github.com/pybind/pybind11/pull/610
2. Change of ``std::vector`` binding style

The first modification is now merged into pybind11 master branch.
The pybind11 in Open3D should be upgraded to mainstream stable pybind11 version.

**Comment by qianyizh** [comment]
Ah, I remember.

I hated the bind_vector function from pybind11, because its ``__repr__`` prints the complete list of all elements. For example, when you call ``print(pcd.points)`` it prints the whole list which can be extremely long.

This is not the behavior I wanted, so I copied the pybind11 convenient function ``bind_vector`` but removed the ``__repr__`` part, and called that function ``bind_vector_without_repr``.

Can you check:

1. If pybind11 still have the long list printing ``__repr__`` behavior? If it is fixed, then we just replace all ``bind_vector_without_repr`` with ``bind_vector``. The ``__repr__`` part should also be removed from ``pybind_eigen_vector_of_vector `` and ``pybind_eigen_vector_of_matrix``.
2. If it is not fixed. Then copy out the body of the new ``bind_vector`` function, paste them into ``bind_vector_without_repr`` but remove the ``__repr__`` part.

This should resolve the issue.

**Comment by syncle** [comment]
Good. I am working on this issue.

**Comment by syncle** [comment]
I decided to keep `bind_vector_without_repr`. Actually the error comes from converting 
`{v.size(), EigenVector::RowsAtCompileTime}` into `detail::any_container<ssize_t>`. It is because pybind2.2 has strict compile time error-check.

This fixes the error:
```
auto vec = py::bind_vector_without_repr<std::vector<EigenVector>>(
			m, bind_name, py::buffer_protocol());
	vec.def_buffer([](std::vector<EigenVector> &v) -> py::buffer_info {
		size_t rows = EigenVector::RowsAtCompileTime;
		return py::buffer_info(
				v.data(), sizeof(Scalar),
				py::format_descriptor<Scalar>::format(),
				2, {v.size(), rows,
				{sizeof(EigenVector), sizeof(Scalar)});
	});
```

Now everything works fine. I tested pybind2.2 with the tutorial scripts.

**Comment by syncle** [comment]
The remaining job is to replace `___init___` for classes. It is not critical issue, but highly recommended by pybind2.2 to avoid runtime warnings.

For example, (from [pybind upgrade guide](http://pybind11.readthedocs.io/en/stable/upgrade.html#v2-2))

```
// old -- deprecated (runtime warning shown only in debug mode)
py::class<Foo>(m, "Foo")
    .def("__init__", [](Foo &self, ...) {
        new (&self) Foo(...); // uses placement-new
    });

// new
py::class<Foo>(m, "Foo")
    .def(py::init([](...) { // Note: no `self` argument
        return new Foo(...); // return by raw pointer
        // or: return std::make_unique<Foo>(...); // return by holder
        // or: return Foo(...); // return by value (move constructor)
    }));
```

Now I am trying to replace 
```
	pinhole_intr.def("__init__", [](PinholeCameraIntrinsic &c, int w, int h,
			double fx, double fy, double cx, double cy) {
		new (&c)PinholeCameraIntrinsic(w, h, fx, fy, cx, cy);
	}, "width"_a, "height"_a, "fx"_a, "fy"_a, "cx"_a, "cy"_a);
```
with 
```
	pinhole_intr.def(py::init([](int w, int h,
	 		double fx, double fy, double cx, double cy) {
	 	return new PinholeCameraIntrinsic(w, h, fx, fy, cx, cy);
	}, "width"_a, "height"_a, "fx"_a, "fy"_a, "cx"_a, "cy"_a));
```
but this makes compile error. It is because of keyword arguments. I can compile it without keyword arguments. @qianyizh do you have any suggestion for this issue?

**Comment by qianyizh** [comment]
What is a keyword argument?

**Comment by syncle** [comment]
It is ``"width"_a, "height"_a, "fx"_a, "fy"_a, "cx"_a, "cy"_a``. If I remove this line, it complies.

**Comment by qianyizh** [comment]
I think they are the decorations to the function, i.e., when you type ``help(xxx.__init__)``, these decorations will pop out as the argument names.

**Comment by qianyizh** [comment]
Addressed in #127
